<a href="https://colab.research.google.com/github/jackiekuen2/notes-handson-ml-tf/blob/master/ch8_ExerciseQ9_PCA_MNIST.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
# a = []
# while(1):
#     a.append('1')

In [0]:
from sklearn.datasets import fetch_openml
import numpy as np

mnist = fetch_openml('mnist_784', version=1)
mnist.target = mnist.target.astype(np.int64)

In [0]:
mnist.data.shape

(70000, 784)

In [0]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(mnist.data, mnist.target, test_size=10000, random_state=42)

# Part A

## Random Forest Classifier #1

In [0]:
from sklearn.ensemble import RandomForestClassifier

forest_clf = RandomForestClassifier(n_estimators=10, random_state=42)

In [0]:
import time

start_time = time.time()

forest_clf.fit(X=X_train, y=y_train)

print("Training time: {}".format(time.time() - start_time))

Training time: 4.348313093185425


In [0]:
from sklearn.metrics import accuracy_score

y_test_pred = forest_clf.predict(X_test)

accuracy_score(y_test, y_test_pred)

0.9465

Random Forest Classifier's testing accuracy: 94.65%

## PCA + Random Forest Classifier #2

In [0]:
from sklearn.decomposition import PCA

pca = PCA(n_components=0.95)
X_train_reduced = pca.fit_transform(X_train)

In [0]:
forest_clf2 = RandomForestClassifier(n_estimators=10, random_state=42)

start_time = time.time()
forest_clf2.fit(X=X_train_reduced, y=y_train)

print("Training time: {}".format(time.time() - start_time))

Training time: 11.788392066955566


The training time with PCA dimensionality reduction is not always faster. It depends on dataset, model and training algorithm.

In [0]:
X_test_reduced = pca.transform(X_test)

y_test_pred = forest_clf2.predict(X_test_reduced)

accuracy_score(y_test, y_test_pred)

0.891

The new Random Forest Classifier (with PCA) testing accuracy: 89.1%

# Part B

## Logistic Regression #1

In [0]:
from sklearn.linear_model import LogisticRegression

log_clf = LogisticRegression(multi_class='multinomial', solver='lbfgs', random_state=42)

start_time = time.time()

log_clf.fit(X_train, y_train)

print("Training time: {}".format(time.time() - start_time))

Training time: 18.81350016593933


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [0]:
y_test_pred = log_clf.predict(X_test)

print("Testing accuracy: {}%".format(accuracy_score(y_test, y_test_pred)*100))

Testing accuracy: 91.89%


## PCA + Logistic Regression #2

In [0]:
log_clf2 = LogisticRegression(multi_class='multinomial', solver='lbfgs', random_state=42)

start_time = time.time()

log_clf2.fit(X_train_reduced, y_train)

print("Training time: {}".format(time.time() - start_time))

Training time: 9.481841564178467


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [0]:
y_test_pred = log_clf2.predict(X_test_reduced)

print("Testing accuracy: {}%".format(accuracy_score(y_test, y_test_pred)*100))

Testing accuracy: 91.41%


A slight drop in testing accurary, but shorten the training time by half